### PR TITLE
[optin/reporting] task: refactor query and validation

### PR DESCRIPTION
**Pull request by AnthonyLaw**
### What was the issue?
- current SQL query is using concatenation, to prevent SQL injection we need to use bind params.
- That is zero validation in req.query 

### What's the fix?
- rename req query `tz` to `timezone`
- refactor on the SQL query with bind params.
- added express-validator package
- added req query validation and error handling.

### CODE REVIEW

File: optin/reporting/controllers/optinRequestsController.js [hunk 1]
```diff
@@ -57,15 +54,15 @@ const controller = {
 			}
 		};
 
-		const status = statusStringToInt(statusFilter.toLowerCase());
+		const statusFilter = statusStringToInt(status.toLowerCase());
```
Comment (Jaguar0625) [inline optin/reporting/controllers/optinRequestsController.js:57]: why did you rename this? if it is single value (int), i think status name is clearer. "filter" typically connotates a function.

File: optin/reporting/utils/Validation.js [hunk 2]
```diff
@@ -0,0 +1,664 @@
+const { query, validationResult } = require('express-validator');
+
+const validMomentTimezones = [
```
Comment (Jaguar0625) [inline optin/reporting/utils/Validation.js:3]: are these not defined in some package?

File: optin/reporting/utils/Validation.js [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+
+const Validation = {
+	validate: method => {
+		const paginationValidation = [
+			query('pageSize', 'pageSize required').exists().isInt(),
+			query('pageNumber', 'pageNumber required').exists().isInt()
+		];
+
+		const sortValidation = [
+			query('sortBy', 'field name must be string')
```
Comment (Jaguar0625) [inline optin/reporting/utils/Validation.js:606]: should you limit possible sortBy values?

File: optin/reporting/utils/Validation.js [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+		if (!errors.isEmpty()) {
+			res.status(422).json({ errors: errors.array() });
+			return;
+		}
+
+		next();
+	}
+};
+module.exports = Validation;
```
Comment (Jaguar0625) [inline optin/reporting/utils/Validation.js:664]: should you add some unit tests for this?

File: optin/reporting/utils/Validation.js (hunk 3 above)
Comment (AnthonyLaw) [reply optin/reporting/utils/Validation.js:606]: Agree, i will add that 👍🏼

File: optin/reporting/utils/Validation.js [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+			];
+		}
+
+		case 'exportCsv': {
+			return [
+				query('timezone', 'timezone must be string example America/Los_Angeles').exists().isIn(validMomentTimezones)
+			];
+		}
+
+		case 'getOptinRequests': {
```
Comment (yilmazbahadir) [inline optin/reporting/utils/Validation.js:639]: I'd move page-specific validations next to `Controller`s rather than a utility file. So `getCompleted` and `getOptinRequests` can go next to (or inside) the controllers and the rest would stay here as shared validationUtility, _if that makes sense_

File: optin/reporting/controllers/optinRequestsController.js (hunk 1 above)
Comment (AnthonyLaw) [reply optin/reporting/controllers/optinRequestsController.js:57]: I rename that because I did destructure params. I will undo this part.

File: optin/reporting/utils/Validation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnthonyLaw) [reply optin/reporting/utils/Validation.js:3]: I will remove the list, it actually can find from the package `moment-timezone`.

File: optin/reporting/utils/Validation.js (hunk 5 above)
Comment (AnthonyLaw) [reply optin/reporting/utils/Validation.js:639]: The reason I put all validation in 1 place, is because it will be easier to manage IMO.

Some of the validation rules are shared between `getCompleted` and `getOptinRequests`.
if we move to page-specific validations `Controllers`, we may have some duplicated code.

**Review by Jaguar0625 — APPROVED** [review]
